### PR TITLE
Attribute access for partially split objects

### LINKED ISSUE #26: Allow attribute access to PartiallyKwargifiedBytes
Stemming from this conversation with @michwill:

https://github.com/nucypher/nucypher/pull/1451#discussion_r361316656

### PR DESCRIPTION
**Pull request by jMyles**
Fixes #26

### CODE REVIEW

File: tests.py [hunk 1]
```diff
@@ -218,32 +218,72 @@ def from_bytes(cls, thing_as_bytes, necessary_kwarg=False):
...
+        self.size = size
 
-    fab = BytestringKwargifier(
-        ThingToBeFabricated,
-        thing1=VariableLengthBytestring,
-        thing2=(bytes, 13),
-        thing3=(int, 2, {"byteorder": "big"})
-    )
+    def sip(self):
+        return "Mmmm"
```
Comment (cygnusv) [inline tests.py:238]: > ELAINE: Yeah, in college I sat next to an Alex in art history. And he was always drinking coffee and after every sip he would go: "Ahh!". I mean every two seconds: "Ahh!". And he would take like 40 sips and after everyone: "Ahh!". I had to drop the class.

File: bytestring_splitter/__init__.py [hunk 2]
```diff
@@ -44,18 +48,33 @@ def set_original_bytes_repr(self, bytes_representation):
...
+            return self._finished_values[message_name]
+        except KeyError:  # suppress might be good here, but it appears to have a performance penalty, and this is a performance-concerned function.
+            pass
+
+        try:
+            bytes_for_message, message_class, kwargs = self.processed_objects[message_name]
+            self._finished_values[message_name] = produce_value(message_class,
+                                                               message_name,
+                                                               bytes_for_message,
+                                                               kwargs)
```
Comment (cygnusv) [inline bytestring_splitter/__init__.py:71]: An extremely minor comment: My obsessive-compulsive self tells me that both calls to `produce_value()` (here and L53) have the arguments not well aligned.

File: bytestring_splitter/__init__.py [hunk 3]
```diff
@@ -44,18 +48,33 @@ def set_original_bytes_repr(self, bytes_representation):
...
+            produced_value = produce_value(message_class, message_name, bytes_for_message, kwargs)
+            del self.processed_objects[message_name]  # We don't do this as a pop() in case produce_value raises.
```
Comment (cygnusv) [inline bytestring_splitter/__init__.py:73]: I don't understand this comment, can you elaborate?

**Review by cygnusv — APPROVED** [review]

**Review by KPrasch — APPROVED** [review]

File: bytestring_splitter/__init__.py [hunk 4]
```diff
@@ -44,18 +48,33 @@ def set_original_bytes_repr(self, bytes_representation):
...
                                                           kwargs)
-        return self._receiver(**finished_values, **self._additional_kwargs)
+        return self._receiver(**self._finished_values, **self._additional_kwargs)
+
+    def __getattr__(self, message_name):
+        # First we'll try to see if this message_name already has a finished value:
+        try:
+            return self._finished_values[message_name]
+        except KeyError:  # suppress might be good here, but it appears to have a performance penalty, and this is a performance-concerned function.
+            pass
```
Comment (KPrasch) [inline bytestring_splitter/__init__.py:64]: Heh - interesting - does `.get` share a similar performance hit?

**Review by vepkenez — APPROVED** [review]
I like it.  Why are we doing this now?  Where is this needed?

**Comment by KPrasch** [comment]
@vepkenez - It is needed in nucypher in order to read the `domain` attribute from a `NodeSprout` without maturing it, subsequently this will help resolve an outstanding issue regarding unintentional cross-domain node discovery. See <https://github.com/nucypher/nucypher/pull/1868>

**Comment by jMyles** [comment]
@vepkenez: It's also something like 15% more optimal on its own (though I'm not doing bechmarking right now except with very quick local `timeit`s).

File: bytestring_splitter/__init__.py (hunk 3 above)
Comment (jMyles) [reply bytestring_splitter/__init__.py:73]: It'd be marginally more optimal to do `self.processed_objects.pop(message_name)`, but in the (rare, but not unsupported) case that something changes in the receiver class such that `produce_value` raises at this time (but is caught and subsequently proceeds successfully), we'd already have erased it from `processed_objects`.

File: bytestring_splitter/__init__.py (hunk 4 above)
Comment (jMyles) [reply bytestring_splitter/__init__.py:64]: `.get` has what appears to be a measurable, but substantially smaller performance hit.  Here's [what I said in Discord](https://ptb.discordapp.com/channels/411401661714792449/411401661714792451/698011920736976937) about it:

---
Interesting: I just ran three test scenarios on the "just in time attribute access", each access an attribute 500000 times from a partially split bytestring.

It's possible that the value has been "finished" (ie, passed to its receiver), so we have to both check that and return the finished value, and then finish the value if it hasn't previously been.

They are both dicts.

I tried with .get(...), with a try block (and except: pass), and suppress.

The times were more different than I thought:

The try block was about 4.2 seconds, the .get was about 4.6 seconds, and the suppress was 6.2 seconds (!).
---

At some point, we can do some more formal benchmarking.  It doesn't feel like the best use of my time right now.

**Pull request merged**